analyzers/src/SonarAnalyzer.CSharp/Rules/BooleanLiteralUnnecessaryCodeFix.cs

```diff
@@ -250,73 +250,66 @@ private static Document RemoveConditional(Document document, SyntaxNode root, Co
 
                 return document.WithSyntaxRoot(newRoot);
             }
 
-            if (whenTrue.Equals(syntaxNode)
-                && CSharpEquivalenceChecker.AreEquivalent(syntaxNode, CSharpSyntaxHelper.FalseLiteralExpression))
+            if (whenTrue.Equals(syntaxNode) && syntaxNode.IsFalse())
             {
                 var newRoot = ReplaceExpressionWithBinary(
                     conditional,
                     root,
                     SyntaxKind.LogicalAndExpression,
                     GetNegatedExpression(conditional.Condition),
-                    conditional.WhenFalse);
+                    ParenthesizeIfComplex(conditional.WhenFalse));
 
                 return document.WithSyntaxRoot(newRoot);
             }
 
             var whenFalse = conditional.WhenFalse.RemoveParentheses();
 
-            if (whenFalse.Equals(syntaxNode)
-                && CSharpEquivalenceChecker.AreEquivalent(syntaxNode, CSharpSyntaxHelper.TrueLiteralExpression))
+            if (whenFalse.Equals(syntaxNode) && syntaxNode.IsTrue())
             {
                 var newRoot = ReplaceExpressionWithBinary(
                     conditional,
                     root,
                     SyntaxKind.LogicalOrExpression,
                     GetNegatedExpression(conditional.Condition),
-                    conditional.WhenTrue);
+                    ParenthesizeIfComplex(conditional.WhenTrue));
 
                 return document.WithSyntaxRoot(newRoot);
             }
 
-            if (whenFalse.Equals(syntaxNode)
-                && CSharpEquivalenceChecker.AreEquivalent(syntaxNode, CSharpSyntaxHelper.FalseLiteralExpression))
+            if (whenFalse.Equals(syntaxNode) && syntaxNode.IsFalse())
             {
                 var newRoot = ReplaceExpressionWithBinary(
                     conditional,
                     root,
                     SyntaxKind.LogicalAndExpression,
                     conditional.Condition,
-                    conditional.WhenTrue);
+                    ParenthesizeIfComplex(conditional.WhenTrue));
 
                 return document.WithSyntaxRoot(newRoot);
             }
 
             return document;
         }
 
-        private static ExpressionSyntax GetNegatedExpression(ExpressionSyntax expression)
-        {
-            var exp = expression;
-            if (expression is BinaryExpressionSyntax
-               || expression is ConditionalExpressionSyntax)
-            {
-                exp = SyntaxFactory.ParenthesizedExpression(expression);
-            }
+        private static ExpressionSyntax GetNegatedExpression(ExpressionSyntax expression) =>
+            SyntaxFactory.PrefixUnaryExpression(SyntaxKind.LogicalNotExpression, ParenthesizeIfComplex(expression));
 
-            return SyntaxFactory.PrefixUnaryExpression(SyntaxKind.LogicalNotExpression, exp);
-        }
+        private static ExpressionSyntax ParenthesizeIfComplex(ExpressionSyntax expression) =>
```

Comment:
Easier: Alyways Parenthsize and add the `Simplifier.Annotation` to the newly created expression. Roslyn should remove the parenthesis if not needed.
See also https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.simplification.simplifier?view=roslyn-dotnet-4.6.0

Reply:
From what I understood, you need to pass to `ReduceAsync` method the full document, marking with `Annotation` the sub-tree which you are interested to reduce. Then I wonder how to retrieve back the same reduced SyntaxNode from the full document, any ideas?

Reply:
Sorry, with `.WithAdditionalAnnotations(Simplifier.Annotation)` it's actually working fine, great!